src/Core/Util/FormChoiceFormatter.php

```diff
@@ -0,0 +1,89 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Util;
+
+/**
+ * Class that formats choices for Symfony forms.
+ */
+class FormChoiceFormatter
+{
+    /**
+     * Returns a choice list in a format required for Symfony form.
+     * Symfony form choice fields accept array with options, where the item name
+     * is the key and item ID is the value.
+     *
+     * So, when there are two items with the same name, they get lost.
+     * This will automatically mark duplicate options with their ID.
+     *
+     * @param array $rawChoices Raw array with data to build the options from
+     * @param string $idKey Key name of the item IDs, id_carrier for example
+     * @param string $nameKey Key name of the item NAMEs, carrier_name for example
+     *
+     * @return array Formatted choices
+     */
+    public static function formatFormChoices(array $rawChoices, string $idKey, string $nameKey): array
+    {
+        // Final array with choices
+        $finalChoices = [];
+
+        // A slim array with just they keys we processed, so we know what are duplicates
+        $alreadyProcessedKeys = [];
+
+        foreach ($rawChoices as $rawChoiceKey => $rawChoice) {
+            // If we already came across this exact value name before, we will
+            // append the option ID before the name.
+            if (in_array($rawChoice[$nameKey], $alreadyProcessedKeys)) {
+                // We store it with ID prepended before the name.
+                $finalChoices[sprintf('%s (%d)', $rawChoice[$nameKey], $rawChoice[$idKey])] = $rawChoice[$idKey];
+
+                // And if it's the first duplicate (second occurence), we also modify the
+                // previous normal one.
+                if (isset($finalChoices[$rawChoice[$nameKey]])) {
+                    $previousId = $finalChoices[$rawChoice[$nameKey]];
+                    unset($finalChoices[$rawChoice[$nameKey]]);
+                    $finalChoices[sprintf('%s (%d)', $rawChoice[$nameKey], $previousId)] = $previousId;
+                }
+            } else {
+                // We store it in the final array normally
+                $finalChoices[$rawChoice[$nameKey]] = $rawChoice[$idKey];
+            }
+
+            // For next time, we mark that we processed this option
+            $alreadyProcessedKeys[] = $rawChoice[$nameKey];
+
+            // And save some memory, we don't need it anymore
+            unset($rawChoices[$rawChoiceKey]);
```

Comment:
Not sure this is needed or if it brings any improvement on performances 🤔 You can simply loop through the array without modifying it If so you can also simplify the for loop:
```php
foreach ($rawChoices as $rawChoice) {
```

Reply:
My reasoning was that I gradually cut from one array and add it, so we don't increase memory usage by looping it. :-)